sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/SparkSQLParser.scala

```diff
@@ -25,15 +25,51 @@ import scala.util.parsing.input.CharArrayReader.EofCh
 
 import org.apache.spark.sql.catalyst.plans.logical._
 
-private[sql] abstract class AbstractSparkSQLParser
+private[sql] trait KeywordNormalizer {
+  def apply(str: String): String = str
+}
+
+private[sql] object CaseInsensitive extends KeywordNormalizer {
+  override def apply(str: String) = str.toLowerCase()
+}
+
+private[sql] object CaseSensitive extends KeywordNormalizer {
+  override def apply(str: String) = str
+}
+
+// TODO make the normalizer configurable?
+private[sql] abstract class AbstractSparkSQLParser(normalizer: KeywordNormalizer = CaseInsensitive)
   extends StandardTokenParsers with PackratParsers {
 
-  def apply(input: String): LogicalPlan = phrase(start)(new lexical.Scanner(input)) match {
-    case Success(plan, _) => plan
-    case failureOrError => sys.error(failureOrError.toString)
+  def apply(input: String): LogicalPlan = {
```

Comment:
You're right, we probably can combine couple of `Parser` work in delegation mode, but currently, I just simply wrote another version of the `def apply` in `DDLParser`.